### PR TITLE
Perf

### PR DESCRIPTION
**Pull request by mredaelli**
A few changes

The result is going from
```
satisfia (b28010a) 📦🤷 🐍 v3.10.13 via ⊢💩 (shell) on ☁️  m.redaelli@gmail.com
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):     38.970 s ±  0.577 s    [User: 38.611 s, System: 0.551 s]
  Range (min … max):   38.037 s … 39.630 s    10 runs
```
to
```
satisfia on 🌱 perf 📦🤷 🐍 v3.10.13 via ⊢💩 (shell) on ☁️  m.redaelli@gmail.com
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):     30.802 s ±  1.359 s    [User: 30.457 s, System: 0.587 s]
  Range (min … max):   29.625 s … 33.938 s    10 runs
```

last commit goes down to 24 secs

### CODE REVIEW

File: src/satisfia/agents/makeMDPAgentSatisfia.py [hunk 1]
```diff
@@ -815,54 +817,57 @@ def randomTieBreaker(self, state, action):
 	# now we can combine all of the above quantities to a combined (safety) loss function:
 
 	# state, action, aleph4state, aleph4action, estActionProbability
-	#@lru_cache(maxsize=None)
+	@lru_cache(maxsize=None)
```
Comment (mredaelli) [inline src/satisfia/agents/makeMDPAgentSatisfia.py]: this was disabled but it seems to help.

was there a reason why you disabled it?

If it was a memory problem, it might be worth to put a limit, like 1024: it would probably still help

File: src/satisfia/agents/makeMDPAgentSatisfia.py [hunk 2]
```diff
@@ -1380,6 +1385,7 @@ def X(nextState):
 			print(pad(state),"| | | | | ╰ ETerminalState2_action", prettyState(state), action, aleph4action, policy, ":", res)
 		return res
 
+	@cache
```
Comment (mredaelli) [inline src/satisfia/agents/makeMDPAgentSatisfia.py]: this also was not cached. Any reason against it?

File: src/satisfia/util/distribution.py [hunk 3]
```diff
@@ -40,7 +38,7 @@ def var(self, *, precision=64):
 		return sum((samples - E) ** 2) / (precision - 1) # sample variance
 
 	def stddev(self, *, precision=64):
-		return torch.sqrt(self.var(precision))
+		return math.sqrt(self.var(precision))
```
Comment (mredaelli) [inline src/satisfia/util/distribution.py:41]: would you mind this? we don't need it yet, and it messes up with browser support

File: src/satisfia/util/helper.py [hunk 4]
```diff
@@ -1,56 +1,42 @@
...
+from typing import NamedTuple
+
+def Interval(left, right=None):
+    if hasattr(left, "__len__"):
+        assert right is None
+        return _Interval(*left)
+    else:
+        return _Interval(left,  left if right is None else right)
+
+class _Interval(NamedTuple):
```
Comment (mredaelli) [inline src/satisfia/util/helper.py:12]: namedtuples are much faster than generic classes.

This change should be transparent to the rest of the code

**Comment by mredaelli** [comment]
Caching also `propagateAspiration` brings it down to 29s, and the memory usage for stays below 450MB.

**Comment by mredaelli** [comment]
Next things to consider would be:

- `_extract_state_attributes` takes a lot of time. If `state` instead were already a tuple, we'd save all that computation plus the time to call the function. Would that be messy?
- `transition_distribution`

File: src/satisfia/agents/makeMDPAgentSatisfia.py (hunk 2 above)
Comment (mensch72) [reply src/satisfia/agents/makeMDPAgentSatisfia.py]: i might simply have forgotten it

File: src/satisfia/util/distribution.py (hunk 3 above)
Comment (mensch72) [reply src/satisfia/util/distribution.py:41]: ok for now, at some point we might however need it for training.
what about 
```
try:
  import torch
  sqrt = torch.sqrt
except:
  sqrt = math.sqrt  # or numpy?
```

File: src/satisfia/util/helper.py (hunk 4 above)
Comment (mensch72) [reply src/satisfia/util/helper.py:12]: very good!

File: src/satisfia/util/distribution.py (hunk 3 above)
Comment (mredaelli) [reply src/satisfia/util/distribution.py:41]: My guess would be that torch can be relegated to files that are imported only from the training part, which in turn are not imported by the web book, so no problem arises.

If that turns out not to be true, yes, your solution should work

**Comment by mredaelli** [comment]
I pushed a cleanup of the utilities, with the interval version and the "scalar but maybe second argument is an interval" version:

```
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):     26.851 s ±  0.742 s    [User: 26.346 s, System: 0.600 s]
  Range (min … max):   26.273 s … 28.715 s    10 runs
```

so we are saving about 1/3 of the time.

I actually specialized the utilities a bit more, adding a scalar version (e.g. `clip`), a version where only the central argument is a tuple (`clip2`) and a generic one for intervals (`clipInt`), the original one, which however doesn't seem to be used.

That saves a bit more, and I like it better because it's more... "typed" :), but I understand if it gets in the way:
```
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
 ⠦ Current estimate: 26.614 s     ████████████████████████████░░░░░░░░░░░░░░░░░░░░ ETA 00:01:46  Time (mean ± σ):     26.436 s ±  0.631 s    [User: 25.798 s, System: 0.607 s]
  Range (min … max):   25.509 s … 27.904 s    10 runs
```

**Comment by mredaelli** [comment]
Latest commit shaves another couple of seconds by excluding from the loss computation stuff with 0 loss parameter directly in `__init__`. 

Losses for time and `lossCoeff4OtherLoss` still need adaptation though.

I also replaces all tabs with spaces because I'm going crazy with whitespace errors. I hope that's OK. To see the actual difference, disable whitespace in the diff config:

image

File: src/satisfia/agents/makeMDPAgentSatisfia.py [hunk 5]
```diff
@@ -17,1546 +17,1522 @@
...
+    # now we can combine all of the above quantities to a combined (safety) loss function:
+
+    # state, action, aleph4state, aleph4action, estActionProbability
+    @lru_cache(maxsize=None)
+    def combinedLoss(self, s, a, al4s, al4a, p): # recursive
+
+        if self.debug:
+            print(pad(s),"| | combinedLoss, state",prettyState(s),"action",a,"aleph4state",al4s,"aleph4action",al4a,"estActionProbability",p,"...")
+
+        res = sum(k * v(s, a, al4s, al4a, p) for k, v in self.losses.items())
```
Comment (mensch72) [inline src/satisfia/agents/makeMDPAgentSatisfia.py]: shouldn't we also make sure that `v` is only called if `k != 0`, like so: `0 if k == 0 else k * v(...)`?

File: src/satisfia/agents/makeMDPAgentSatisfia.py [hunk 6]
```diff
@@ -17,1546 +17,1522 @@
...
+
+        if self.verbose or self.debug:
+            print(pad(s),"| | combinedLoss, state",prettyState(s),"action",a,"aleph4state",al4s,"aleph4action",al4a,"estActionProbability",p,":",res,"\n"+pad(s),"| |    ")
+
```
Comment (mensch72) [inline src/satisfia/agents/makeMDPAgentSatisfia.py:874]: I think we should print the nonzero loss components at least if `debug==True`

File: src/satisfia/util/helper.py [hunk 7]
```diff
@@ -1,84 +1,81 @@
...
-		return 2
+def interpolateInt(x, l, y):
+    # one argument is an interval, so everything becomes an interval
+    x, l, y = Interval(x), Interval(l), Interval(y)
+    return Interval(x[0] + l[0] * (y[0] - x[0]), x[1] + l[1] * (y[1] - x[1]))
 
-def interpolate(x, l, y):
-	# denoted	x : l : y in formulas
+def relativePosition2(x, z, y):
+    if (diff := (y-x)) == 0.:
```
Comment (mensch72) [inline src/satisfia/util/helper.py:51]: Interesting, I wasn't aware of the `:=` operator in Python. Good to know...

File: src/world_model/world_model.py [hunk 8]
```diff
@@ -62,27 +62,26 @@ def default_policy(self, state):
         return None
 
     @lru_cache(maxsize=None)
-    def possible_successors(self, state, action=None, n_samples = None):
-        """Return a list of possible successor states after performing action in state,
+    def possible_successors(self, state, action=None, n_samples = None) -> set:
+        """Return the of possible successor states after performing action in state,
```
Comment (mensch72) [inline src/world_model/world_model.py:77]: typo, it should be "Return the set of ..."

**Review by mensch72 — APPROVED** [review]
all fine apparently as judged from code review and regression test.

**Comment by mensch72** [comment]
Hopefully the merge is not too difficult...

File: src/satisfia/util/helper.py (hunk 7 above)
Comment (simfis) [reply src/satisfia/util/helper.py:51]: It's pretty new: https://peps.python.org/pep-0572/

File: src/satisfia/agents/makeMDPAgentSatisfia.py (hunk 5 above)
Comment (mredaelli) [reply src/satisfia/agents/makeMDPAgentSatisfia.py]: If I'm not mistaken, that's done in `__init__`, here:

https://github.com/pik-gane/satisfia/pull/7/files#diff-2a77c67961414a5d12c1a7db8c3b4536e945a06643c4f16f68f3509b002378eeR190

so we do it only once rather than at every evaluation.

Unless the parameters can be changed?

File: src/satisfia/agents/makeMDPAgentSatisfia.py (hunk 6 above)
Comment (mredaelli) [reply src/satisfia/agents/makeMDPAgentSatisfia.py:874]: True, that needs to be done. If you agree with the rest, I'll add it

**Comment by mensch72** [comment]
parameters cannot be changed during the lifetime of the agent at the moment.

Am 9. April 2024 20:17:52 MESZ schrieb Massimo Redaelli ***@***.***>:

File: src/world_model/world_model.py (hunk 8 above)
Comment (simfis) [reply src/world_model/world_model.py:77]: Fixed.

**Comment by mredaelli** [comment]
rebased and removed some numpy.

Down to:
```
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):     22.559 s ±  0.579 s    [User: 21.415 s, System: 1.334 s]
  Range (min … max):   21.661 s … 23.575 s    10 runs
```

**Comment by mredaelli** [comment]
Added a commit to remove `prev_loc` from state.

```
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):      5.990 s ±  0.359 s    [User: 5.750 s, System: 0.521 s]
  Range (min … max):    5.670 s …  6.830 s    10 runs
```
